kong/clustering.lua

```diff
@@ -343,27 +347,63 @@ local function should_send_config_update(node_version, node_plugins)
                   "automatically once this DP also upgrades to 2.3 or later"
   end
 
-  local minor_cp = KONG_VERSION:match(MAJOR_MINOR_PATTERN)
-  local minor_node = node_version:match(MAJOR_MINOR_PATTERN)
-  if minor_cp ~= minor_node then
-    return false, "version mismatches, CP version: " .. minor_cp ..
-                  " DP version: " .. minor_node,
+  local major_cp, minor_cp = KONG_VERSION:match(MAJOR_MINOR_PATTERN)
+  local major_node, minor_node = node_version:match(MAJOR_MINOR_PATTERN)
+  minor_cp = minor_cp and tonumber(minor_cp) or INF
+  minor_node = minor_node and tonumber(minor_node) or -INF
+
+  if major_cp ~= major_node or minor_cp - 2 > minor_node or minor_cp < minor_node then
+    return false, "version incompatible, CP version: " .. KONG_VERSION ..
+                  " DP version: " .. node_version ..
+                  " DP versions acceptable are " ..
+                  major_cp .. "." .. math.max(0, minor_cp - 2) .. " to " ..
+                  major_cp .. "." .. minor_cp .. "(edges included)",
                   CLUSTERING_SYNC_STATUS.KONG_VERSION_INCOMPATIBLE
   end
 
-  for i, p in ipairs(PLUGINS_LIST) do
-    local np = node_plugins[i]
+  -- allow DP to have a superset of CP's plugins
+  local p, np
+  local i, j = #PLUGINS_LIST, #node_plugins
+
+  if j < i then
+    return false, "CP and DP does not have same set of plugins installed",
+                  CLUSTERING_SYNC_STATUS.PLUGIN_SET_INCOMPATIBLE
+  end
+
+  while i > 0 and j > 0 do
+    p = PLUGINS_LIST[i]
+    np = node_plugins[j]
+
     if p.name ~= np.name then
-      return false, "CP and DP does not have same set of plugins installed",
-                    CLUSTERING_SYNC_STATUS.PLUGIN_SET_INCOMPATIBLE
+      goto continue
     end
 
-    if p.version ~= np.version then
-      return false, "plugin \"" .. p.name .. "\" version differs between " ..
-                    "CP and DP, CP has version " .. tostring(p.version) ..
-                    " while DP has version " .. tostring(np.version),
+    -- ignore plugins without a version (route-by-header is deprecated)
+    if p.version and np.version then
+      local major_p, minor_p = p.version:match(MAJOR_MINOR_PATTERN_PLUGIN)
+      local major_np, minor_np = np.version:match(MAJOR_MINOR_PATTERN_PLUGIN)
+      minor_p = minor_p and tonumber(minor_p) or INF
+      minor_np = minor_np and tonumber(minor_np) or -INF
+
+      if major_p ~= major_np or minor_p - 2 > minor_np or minor_p < minor_np then
```

Comment:
I may forgot to think about the case of plugins in the discussion. Does supporting two minor version ahead for plugins actually makes sense? Since plugins are usually poorly versioned anyway and generally has a slower release pace than core, two minor versions behind may be a lot of changes.

Ping @PaulFischer-Kong as well.

Reply:
This is now updated to only compare plugins's minor version (same as EE currently does)